### PR TITLE
Override response format

### LINKED ISSUE #102: Add the responseFormat option
It should be `'blob' | 'text' | 'json'` and should have higher priority than the response content-type detection

### PR DESCRIPTION
**Pull request by emirotin**
Fixes #102

### CODE REVIEW

File: lib/utils.coffee [hunk 1]
```diff
@@ -274,10 +276,16 @@ exports.getBody = processBody = (response) ->
 				return response.buffer()
 			throw new Error('This `fetch` implementation does not support decoding binary streams.')
 
-		if includes(contentType, 'application/json')
+		if responseFormat is 'json' or (not responseFormat? and includes(contentType, 'application/json'))
 			return response.json()
 
-		return response.text()
+		if responseFormat is 'none'
```
Comment (thgreasi) [inline lib/utils.coffee]: How about moving this `if` first, to make the code more symmetric?

File: lib/request.coffee [hunk 2]
```diff
@@ -172,6 +174,8 @@ module.exports = getRequest = ({
 	# .get('body')
 	###
 	exports.send = (options = {}) ->
+		{ responseFormat } = options
+		delete options.responseFormat
```
Comment (pimterry) [inline lib/request.coffee]: It's not safe to mutate `options` here.

`options` is attached to the ResinRequestError below, if either the request or response fails. Downstream, the UI catches these errors, and retries requests by calling `.send` using the options from the error. With this change, if the UI ever retries a request, it'll drop the `responseFormat` for the retry, and create some fun bugs.

File: lib/utils.coffee [hunk 3]
```diff
@@ -274,10 +276,16 @@ exports.getBody = processBody = (response) ->
...
+			return Promise.resolve(null)
+
+		if not responseFormat? or responseFormat is 'text'
+			return response.text()
+
+		return Promise.reject(new Error("Unknown responseFormat: #{responseFormat}"))
```
Comment (pimterry) [inline lib/utils.coffee]: We now have a `ResinInvalidParameterError`, which would fit here nicely: https://github.com/resin-io-modules/resin-errors/blob/master/lib/errors.coffee#L234

File: tests/response-format.spec.coffee [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+					reader.addEventListener 'loadend', ->
+						resolve(reader.result)
+					reader.readAsText(body)
+			m.chai.expect(promise).to.eventually.satisfy (body) ->
+				s = JSON.stringify(RESPONSE_BODY)
+				if IS_BROWSER
+					return body is s
+				else
+					b = new Buffer(s, 'utf-8')
+					return b.compare(body) is 0
```
Comment (pimterry) [inline tests/response-format.spec.coffee]: Nice tests! :+1:

File: lib/request.coffee (hunk 2 above)
Comment (emirotin) [inline lib/request.coffee]: We actually don't need to delete this custom prop as we fully own the options and whitelist them before passing down to `fetch`

**Comment by emirotin** [comment]
Updated

**Comment by emirotin** [comment]
Build consistently failing on AppVeyor even though all tests pass

**Pull request merged**